File: web/karma_manager.js
class StringCounter {
    constructor() {
        this.counters = new Map();
    }

    countString(c) {
        if ( this.counters.has(c) ) {
            this.counters.set(c, this.counters.get(c) + 1);
        } else {
            this.counters.set(c, 1);
        }
    }

    get(c) {
        if ( this.counters.has(c) ) {
            return this.counters.get(c);
        } else {
            return 0;
        }
    }

    remove(other) {
        var retval = new StringCounter();
        this.counters.forEach(function(value,key){
            var newval = value - other.get(key);
            if ( newval < 0 ) {
                console.log("Tried to remove more than was there in StringCounter");
            } else if ( newval > 0 ) {
                retval.counters.set(key, newval);
            }
        });
        return retval;
    }

    empty() {
        for ( let c of this.counters.keys() ) {
            if ( this.counters.get(c) > 0 ) {
                return false;
            }
        }
        return true;
    }

    subsetOf(other) {
        for ( let key of this.counters.keys() ) {
            if ( other.get(key) < this.get(key) ) {
                return false;
            }
        }
        return true;
    }
}

class KarmaManager {
    constructor() {
        this.dictionary = new Map();
        this.exclusionSet = new Set();
        this.seedSet = [];
        this.pool = [];
    }

    readDictionary(dict) {
        for ( let word of dict ) {
            this.dictionary.set(word, parseWord(word.toLowerCase()));
        } 
    }

    setExclusionSet(ex_set) {
        for ( let word of ex_set ) {
            this.exclusionSet.add(word);
        }
    }
    
    setSeedSet(seed_phrase) {
        this.seedSet.push(seed_phrase);
    }

    clearSeedSet() {
        this.seedSet = [];
    }

    clearExclusionSet() {
        this.exclusionSet.clear();
    }

    processExclusionSet() {
        this.pool = this.pool.filter((word) => { return !this.exclusionSet.has(word) });
    }

    parseTarget(words) {
        this.target = parseWord(words.join(""));
        return this.target;
    }

    findWordPool() {
        this.pool = findSubsets(this.target, this.dictionary.keys(), this.dictionary);
        this.pool.sort(function(a,b){return a.length-b.length || (a>=b?1:-1)})

        return this.pool;
    }

    makeAnagrams(words) {
        this.parseTarget(words);

        this.findWordPool();

        this.processExclusionSet();

        return findCompleteSubsets(this.target, [...this.pool], this.dictionary, this.seedSet);
    }
}

function parseWord(word) {
    let cntr = new StringCounter();
    let new_word = word.toLowerCase();
    for (let index = 0; index < new_word.length; index++ ) {
        var c = new_word[index];
        if ( c >= 'a' && c <= 'z') {
            cntr.countString(c);
        }
    }
    return cntr;
}

let findSubsets = function(target, pool, dictionary) {
    let retval = [];
    for ( let word of pool ) {
        if ( dictionary.get(word).subsetOf(target) ) {
            retval.push(word);
        }
    }
    return retval;
};

function* findCompleteSubsets(target, pool, dictionary, seedset=[], depth=0) {
    if ( seedset.length > 0) {
        for ( let seed of seedset ) {
            newtarget = target.remove(parseWord(seed.join("")));
            if ( newtarget.empty() ) {
                yield seed;
            } else {
                newpool = findSubsets(newtarget, pool, dictionary);
                for ( let subset of findCompleteSubsets(newtarget, newpool, dictionary, [], depth+1) ) {
                    yield seed.concat(subset);
                }
            }
        }
    } else {
        while ( pool.length > 0 ) {
            let word = pool[pool.length - 1];
            // console.log(depth);
            // console.log(word);
            let remainder = target.remove(dictionary.get(word));
            // console.log(remainder.counters);
            if ( remainder.empty() ) {
                yield [word];
            } else {
                for ( let subset of findCompleteSubsets(remainder, findSubsets(remainder, pool, dictionary), dictionary, depth+1) ) {
                    yield [word].concat(subset);
                }
            }
            pool.pop()
        }
    }
}

File: python/karma_manager.py
#!/usr/bin/env python3
# vim:sw=4,ai,expandtab

import logging


class CharCounter(object):
    def __init__(self):
        self._map = {}

    def count(self, c):
        if c in self._map:
            self._map[c] += 1
        else:
            self._map[c] = 1

    def remove(self, other):
        retval = CharCounter()
        remaining = 0
        for c, cnt in self._map.items():
            if c in other._map:
                newtot = cnt - other._map[c]
                if newtot < 0:
                    raise ValueError('Other CharCounter not subset',)
                elif newtot > 0:
                    retval._map[c] = newtot
                remaining += newtot
            else:
                retval._map[c] = cnt
                remaining += cnt
        if remaining > 0:
            return retval
        else:
            return None

    def __le__(self, other):
        for c in self._map.keys():
            if self._map[c] > 0:
                if not (c in other._map and self._map[c] <= other._map[c]):
                    return False
        return True

    def __str__(self):
        s = ''
        chars = list(self._map.keys())
        chars.sort()
        for c in chars:
            if self._map[c] > 0:
                s += c * self._map[c]
        return s
        

class KarmaManager(object):
    def __init__(self, dictionary_file):
        self.dictionary = {}
        with open(dictionary_file, 'r') as f:
            self.readDictionary(f)

    def readDictionary(self, f):
        for word in f.readlines():
            self.addDictionaryWord(word)

    def addDictionaryWord(self, word):
        word = self.smash(word)
        if word not in self.dictionary:
            self.dictionary[word] = self.parseWord(word)
        else:
            logging.warning('Attempted to add duplicate word {} to the dictionary'.format(word))

    def parseWord(self, word):
        retval = CharCounter()
        for c in word:
            if c.isalpha():
                retval.count(c.lower())
        return retval

    def smash(self, word):
        return word.lower().translate(word.maketrans('','','\t\n\r'))

    def makeAnagrams(self, words):
        word = ' '.join(words)
        source = self.parseWord(word)
        logging.debug('Starting with phrase "{}" and source set {}'.format(word, str(source)))

        # This will find only those dictionary words that are a subset of the characters in the
        # input phrase/word.  If the input doesn't have a 'z', then there's no point in including
        # words with 'z' in them.
        working_pool = self._findSubsets(source, self.dictionary.keys())

        # This is a hackish optimization.  It's sorts the words first by length then alphabetically
        # if the same length.  Since _fineCompleteSubsets works backwards through the pool it will
        # start with the longest words first, and work its way backward through the alphabet
        # This will mean that it will find matches with longer words first, which is means quicker
        # results for first match, and that the words it finds will be more interesting.  There
        # is probbably a better optimization out there.
        working_pool.sort(key=lambda x: (len(x), x))

        logging.debug('Starting with {} words total'.format(len(working_pool)))

        return self._findCompleteSubsets(source, working_pool)

    def _findSubsets(self, target, words):
        subsets = []
        for word in words:
            if self.dictionary[word] <= target:
                subsets.append(word)
        return subsets

    def _findCompleteSubsets(self, source, pool, depth=0):

        if not pool:
            return

        while len(pool) > 0:
            word = pool[-1]
            remainder = source.remove(self.dictionary[word])
            logging.debug('{}Find subsets with word {} with pool of {} and remainder {}'.format(' ' * depth, word, len(pool), str(remainder)))
            if remainder is None:
                yield [word,]
            else:
                for subset in self._findCompleteSubsets(remainder, self._findSubsets(remainder, pool), depth+1):
                    yield [word,] + subset
            pool.pop()


def main():
    import argparse
    import sys

    parser = argparse.ArgumentParser('karma_manager', description='anagram maker')
    parser.add_argument('--dictionary', default='dict.txt', help='dictionary to use')
    parser.add_argument('--debug', action='store_true', help='enable debugging output on stderr')
    parser.add_argument('words', type=str, nargs='+', help='words to create anagrams from')

    args = parser.parse_args()

    if args.debug:
        log_level = logging.DEBUG
    else:
        log_level = logging.INFO

    logging.basicConfig(level=log_level)

    km = KarmaManager(args.dictionary)

    anagrams = km.makeAnagrams(args.words)
    counter = 0
    for anagram in anagrams:
        print(' '.join(anagram).capitalize())
        counter += 1;
    logging.info("{} total anagrams found".format(counter))
    if counter > 0:
        sys.exit(0)
    else:
        print('Unable to find anything')
        sys.exit(1)

if __name__ == '__main__':
    main()
